clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -2445,6 +2445,46 @@ static bool isTypeSubstitutable(Qualifiers Quals, const Type *Ty,
   return true;
 }
 
+namespace {
+using DeclContextDesc = std::pair<clang::Decl::Kind, StringRef>;
+bool matchQualifiedTypeName(const QualType &Ty,
+                            ArrayRef<DeclContextDesc> Scopes) {
+  // The idea: check the declaration context chain starting from the type
+  // itself. At each step check the context is of expected kind
+  // (namespace) and name.
+  const CXXRecordDecl *RecTy = Ty->getAsCXXRecordDecl();
+
+  if (!RecTy)
+    return false; // only classes/structs supported
+  const auto *Ctx = dyn_cast<DeclContext>(RecTy);
+  StringRef Name = "";
+
+  for (const auto &Scope : llvm::reverse(Scopes)) {
+    clang::Decl::Kind DK = Ctx->getDeclKind();
+
+    if (DK != Scope.first)
+      return false;
+
+    switch (DK) {
+    case clang::Decl::Kind::ClassTemplateSpecialization:
+      // ClassTemplateSpecializationDecl inherits from CXXRecordDecl
+    case clang::Decl::Kind::CXXRecord:
+      Name = cast<CXXRecordDecl>(Ctx)->getName();
+      break;
+    case clang::Decl::Kind::Namespace:
+      Name = cast<NamespaceDecl>(Ctx)->getName();
+      break;
+    default:
+      llvm_unreachable("matchQualifiedTypeName: decl kind not supported");
+    }
+    if (Name != Scope.second)
+      return false;
+    Ctx = Ctx->getParent();
```

Comment:
Check out getEnclosingNamespaceContext. If we can get the chart to tell us what we expect out of everything (and I think all are namespaces?) you can probably just use that until you find the TU.

Reply:
`getEnclosingNamespaceContext` doesn't seem to work properly:

For `getParent()`, I see the following chain of `(DeclKind, Name)` pairs being analyzed:

```
33      half
14      half_impl
14      detail
14      sycl
14      cl
```

While for `getEnclosingNamespaceContext()` it looks like:

```
33      half
14      half_impl
14      half_impl
```

It seems like enclosing namespace context for `half_impl` is `half_impl`, which is confusing. Probably I don't fully understand something

Reply:
Hmm... interesting.  My reading of the function doesn't really seem like it SHOULD do that, but it is perhaps an old enough function that it isn't terribly maintained.  It just seemed to fit the need :) 

Is that the entire chain? or did you 'give up' there.  There is an interesting call to 'getPrimaryContext' in that function that seems like it should make sure you don't get duplicates...

Reply:
Actually... looking at the logic to that function I think it (or my interpretation of it) is wrong...   I think it would return half_impl forever.  It seems that it never returns the parent of a namespace, just the primary declcontext for the current one (or the namespace containing a current object).  

I looked at some other things that do similar work, so I now think getParent is the only way to do this.